classes/helper.php

```diff
@@ -0,0 +1,193 @@
+
+            $searchsql = $DB->sql_like($columnname, '?', false);
+            $searchparam = '%'.$DB->sql_like_escape($search).'%';
+
+            $sql = "SELECT id, $columnname
+                      FROM {".$table."}
+                     WHERE $searchsql";
+
+            if ($column->meta_type === 'X' || $column->meta_type === 'C') {
+                $records = $DB->get_records_sql($sql, [$searchparam], 0, $limit);
+                if ($records) {
+                    $results[$table][$column->name] = $records;
+                }
+            }
+        }
+
+        return $results;
+    }
+
+    /**
+     * Perform a regular expression search on a table and column.
+     * This function is only called if the database supports regular expression searches.
+     *
+     * @param string $search The regular expression to search for.
+     * @param string $table The table to search.
+     * @param string $column The column to search.
+     * @param $limit The maximum number of results to return.
+     * @return array
+     */
+    private static function regular_expression_search(string $search, string $table,
+                                                      string $column = self::ALL_COLUMNS, $limit = 0): array {
+        global $DB;
+
+        // Check if the database supports regular expression searches.
+        if (!$DB->sql_regex_supported()) {
+            throw new moodle_exception(get_string('errorregexnotsupported', 'tool_advancedreplace'));
+        }
+
+        $results = [];
+
+        $columns = $DB->get_columns($table);
+
+        if ($column !== self::ALL_COLUMNS) {
+            // Only search the specified column.
+            $columns = array_filter($columns, function($col) use ($column) {
+                return $col->name == $column;
+            });
+        }
+
+        foreach ($columns as $column) {
+            $columnname = $DB->get_manager()->generator->getEncQuoted($column->name);
+
+            $select = $columnname . ' ' . $DB->sql_regex() . ' :pattern ';
+            $params = ['pattern' => $search];
+
+            if ($column->meta_type === 'X' || $column->meta_type === 'C') {
+                $records = $DB->get_records_select($table, $select, $params, '', '*', 0, $limit);
+
+                if ($records) {
+                    $results[$table][$column->name] = $records;
+                }
+            }
+        }
+
+        return $results;
+    }
+
+    /**
+     * Perform a search on a table and column.
+     *
+     * @param string $search The text to search for.
+     * @param bool $regex Whether to use regular expression search.
+     * @param string $tables A comma separated list of tables and columns to search.
+     * @param int $limit The maximum number of results to return.
+     * @return array
+     */
+    public static function search(string $search, bool $regex = false, string $tables = '', int $limit = 0): array {
```

Comment:
when we scale up I'm pretty sure this will start breaking as everything is in memory

So probably fine to merge right now, but we'll need to refactor this later. We will probably want to do things in two phases, the first phase returns a list of tables we will search, and then we'll want to do a search table by tables and that second layer will return a record set we can stream through

Reply:
also thinking ahead, we talked about this script just outputting to stdout, but it might be better if it outputs to a file and then we can output a set of progress bars to stdout. Knowing how long things will take and where we are up to with ETA's is going to be fairly important. Again not critical now but I'll make issues for this